overlord/snapstate/snapstate_update_test.go

```diff
@@ -8761,6 +8770,241 @@ func failAfterLinkSnap(ol *overlord.Overlord, chg *state.Change) error {
+				break
+			}
+
+			// Get previous task-set, for the first entry this would be the task-set
+			// in the last entry.
+			var prevTs *state.TaskSet
+			if i == 0 {
+				prevTs = lastTs
+			} else {
+				prevTs = tsByOrder[order[i-1]]
+			}
+			currentTs := tsByOrder[sn]
+
+			// "auto-connect" must depend on "link-snap" of previous
+			acTaskOfCurrent, err := currentTs.Edge(snapstate.MaybeRebootWaitEdge)
+			c.Assert(err, IsNil)
+			linkSnapOfPrev, err := prevTs.Edge(snapstate.MaybeRebootEdge)
+			c.Assert(err, IsNil)
+			c.Check(acTaskOfCurrent.WaitTasks(), testutil.Contains, linkSnapOfPrev)
+		}
+	} else {
+		// Verify by using edges that tasks are correctly connected to the previous
+		// task-set.
+		for i, sn := range order {
+			if i == 0 {
+				continue
+			}
+
+			prevTs := tsByOrder[order[i-1]]
+			currentTs := tsByOrder[sn]
+
+			firstTaskOfCurrent, err := currentTs.Edge(snapstate.BeginEdge)
+			c.Assert(err, IsNil)
+			lastTaskOfPrev, err := prevTs.Edge(snapstate.EndEdge)
+			c.Assert(err, IsNil)
+			c.Check(firstTaskOfCurrent.WaitTasks(), testutil.Contains, lastTaskOfPrev)
+		}
+	}
+
+	// determine the number of reboots we expect
+	s.fakeBackend.linkSnapRebootFor = make(map[string]bool)
+	for _, o := range order {
+		switch o {
+		case "core18", "kernel", "gadget":
+			s.fakeBackend.linkSnapRebootFor[o] = true
+		}
+	}
+	s.fakeBackend.linkSnapMaybeReboot = len(s.fakeBackend.linkSnapRebootFor) > 0
+
+	defer s.se.Stop()
+	s.settle(c)
+
+	if !s.fakeBackend.linkSnapMaybeReboot {
+		// no reboot expected, skip to next checks
+	} else if singleReboot {
+		c.Check(chg.IsReady(), Equals, false)
+		c.Check(chg.Status(), Equals, state.WaitStatus)
+
+		// one reboot expected
+		s.mockRestartAndSettle(c, chg)
+	} else {
+		// perform a reboot for each snap that requires a reboot
+		for _, o := range order {
+			if s.fakeBackend.linkSnapRebootFor[o] {
+				// expect change to be in wait-state before the reboot
+				c.Check(chg.IsReady(), Equals, false)
+				c.Check(chg.Status(), Equals, state.WaitStatus)
+
+				s.mockRestartAndSettle(c, chg)
+			}
+		}
+	}
+
+	c.Check(chg.IsReady(), Equals, true)
+	c.Check(chg.Status(), Equals, state.DoneStatus)
+}
+
+func (s *snapmgrTestSuite) TestUpdateEssentialSnapsOrderAll(c *C) {
+	s.testUpdateEssentialSnapsOrder(c, []string{"snapd", "core18", "gadget", "kernel"}, false)
+}
+
+func (s *snapmgrTestSuite) TestUpdateEssentialSnapsOrderSnapdBase(c *C) {
+	s.testUpdateEssentialSnapsOrder(c, []string{"snapd", "core18"}, false)
+}
+
+func (s *snapmgrTestSuite) TestUpdateEssentialSnapsOrderBaseGadgetKernel(c *C) {
+	s.testUpdateEssentialSnapsOrder(c, []string{"core18", "gadget", "kernel"}, false)
+}
+
+func (s *snapmgrTestSuite) TestUpdateEssentialSnapsOrderBaseKernel(c *C) {
+	s.testUpdateEssentialSnapsOrder(c, []string{"core18", "kernel"}, true)
+}
+
+func (s *snapmgrTestSuite) TestUpdateEssentialSnapsOrderBaseGadget(c *C) {
+	s.testUpdateEssentialSnapsOrder(c, []string{"core18", "gadget"}, false)
+}
```

Comment:
shouldn't we mix these scenarios with bases and apps as well?

Reply:
also from a different angle do we have a test that checks that when we have some non boot bases and some apps with updating base or not, those do not have dependency across each other?

```
some-base---\
  |     some-app-2    some-app-3 (no base update for 3)
some-app-1
```

Reply:
added a test for this case in the reboot suite `TestArrangeSnapTaskSetsLinkageForSnapWithBaseAndWithout`

Reply:
And also added more cases as per your first comment to cover a more broad mix of snap scenarioes.

Reply:
why to the reboot suite and not here?